### PR TITLE
Convert Atmel sam3x series MCU to use ASF

### PR DESCRIPTION
**Pull request by mnkp**
This patch converts Atmel sam3x MCU series to use register header files from Atmel Software Framework (ASF) library. By using ASF different Atmel SAM MCU series can use common device drivers. Deprecation of duplicate drivers will be done in subsequent patches. Minimum set of changes was performed to support ASF, additional work will be necessary to fully integrate sam3x into Atmel SAM family tree.

### CODE REVIEW

File: boards/arm/arduino_due/board.h
Comment (galak) [inline boards/arm/arduino_due/board.h]: this should probably be done in soc.h, not here.

(should probably also send a patch to fixup SAME70 to do the same thing

File: arch/arm/soc/atmel_sam/common/soc_gpio.c
Comment (tbursztyka) [inline arch/arm/soc/atmel_sam/common/soc_gpio.c]: Can't you just do that check at the beginning of the file, once, instead? (if !defined(...) && !defined(...) #error "..." #endif )
Here it's repeating #error and makes the code quite ugly.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: arch/arm/soc/atmel_sam/common/soc_gpio.c
Comment (mnkp) [inline arch/arm/soc/atmel_sam/common/soc_gpio.c]: I'll add the following code at the beginning of the file and remove remaining #error messages.
```
#if    (!defined(PIO_IFSCER_P0) && !defined(PIO_DIFSR_P0)) \
    || (!defined(PIO_IFSCDR_P0) && !defined(PIO_SCIFSR_P0)) \ 
    || (!defined(PIO_ABCDSR_P0) && !defined(PIO_ABSR_P0)) 
#error "Unsupported Atmel SAM MCU series"
#endif
```

File: boards/arm/arduino_due/board.h
Comment (mnkp) [inline boards/arm/arduino_due/board.h]: Good point, I'll make the change (also for SAME70).

**Comment by fallrisk** [comment]
Piotr, do you have an Arduino Due? Did you compile and test this?

**Comment by fallrisk** [comment]
Is this from the Atmel Software Framework (Atmel) or the Advanced Software Framework (Microchip). I don't know much about licenses, but has anyone confirmed the licenses? I'm attempting to download both frameworks and see what their licenses are.

**Comment by mnkp** [comment]
I pushed one more version to remove merge conflict in boards/arm/arduino_due/arduino_due_defconfig

**Review by fallrisk — COMMENTED** [review]
I am looking into what version of ASF we have and if it came from Atmel or Microchip and looking into the licenses. If someone determines this information before me post it here.

**Comment by mnkp** [comment]
The ASF files for SAM3X live in ext/hal/atmel/asf/sam/include/sam3x/
They were introduced in Zephyr by commit b54f2a4. The commit message states origin and the license.

**Comment by fallrisk** [comment]
We should rename the the directory "ext/hal/atmel/asf" to "ext/hal/atmel/dfp" because the files you keep importing are from http://packs.download.atmel.com/ and are called device family packs. These aren't part of ASF. These device family packs when downloaded show an Apache 2.0 license and the individual files all have the 2 part BSD license header. @carlescufi said that the 2-part BSD license was OK for us to use.

**Comment by mnkp** [comment]
Thanks @fallrisk for investigating the license issue!

The files were placed in asf/ directory as originally the code come from ASF. Later we got SAME70 headers with Apache 2.0 license, those were from device family packs. Also at the time it was not clear, it's in fact still not clear, if apart from header files we are going to use any source code files. Source code files are available in ASF only. I'm generally in favor of not using any ASF source code as its quality is low, but if someone needs to add quickly let's say USB support using ASF may be the best course of action. Thus, we probably shouldn't exclude this possibility.

My original suggestion when adding SAM E70 header files was not to use ASF name or asf/ folder at all but rather place all the files in ext/hal/atmel/. Vendors tend to change names of their support libraries every couple of years and we would also avoid problems like the one we're just discussing.

I'm not convinced we should rename asf/ directory just yet, but I will do as requested. In any case I propose to merge this commit as it is, assuming there are no other issues, and fix header files location later.

**Comment by fallrisk** [comment]
Don't change it unless one of the main people agree with me. I apologize for it looking like I was wanting you to change that directory name. I want the managers to look at what I found and make the determination. Right now your PR looks good to me and I will approve it as soon as one of the managers just looks at my consideration.

**Comment by mnkp** [comment]
Hi @fallrisk, no need to apologize. The issue you raised is certainly worth discussing. You've presented your point of view, I presented mine. At the end we all (including maintainers), hopefully come to some fruitful conclusion.

It's quite unfortunate that in open source project like Zephyr getting acquainted with a fellow coder starts with reviews. Pointing out their mistakes and differences in opinion. Nevertheless, from what I've seen so far, we are all friendly crowd.

File: arch/arm/soc/atmel_sam/common/soc_gpio.c
Comment (mnkp) [inline arch/arm/soc/atmel_sam/common/soc_gpio.c]: Hi @tbursztyka, I've provided a fix for the issue you reported. Is there anything else I should change?

**Comment by galak** [comment]
We need to get Atmel to release updated SDK files with apache license.  We probably should never have allowed the import of the SAM3X files as the license isn't technically acceptable to Zephyr.

So until we resolve that (I have email requests into the Atmel / Microchip guys).  This should not be merged.

**Comment by galak** [comment]
Can you rebase this and push against the 'arm' branch.

**Comment by galak** [comment]
My previous comments about license can be ignored, didnt realize that the pack ASF files are just BSD style licensed so they are ok.

**Comment by mnkp** [comment]
Great, I'm happy to hear the license issue was cleared!

I can push this PR against 'arm' branch, no problem. I would like to ask however if this is a one time/special request or I should push all future Atmel SAM related commits against 'arm'? Following the email "Move to Github completed and some issue" from @nashif  I was convinced we don't work with subsystem branches anymore and all push requests should be made against master. 'arm' branch is currently the only one where there is still some development going on, everyone else is pushing to master.

**Comment by galak** [comment]
For the arm work I want to keep doing things on a branch, so this isn't a one time request, but going forward.  We will work to improve the process associated with branches, but will still require from time to time rebases.

**Comment by mnkp** [comment]
I have rebased the commits on 'arm' branch and pushed to github but it looks like it does not work as expected. Github still wants to merge it into master, along some 20 or so unrelated commits. I've since rebased it back against master and re-pushed. Maybe we can merge it exceptionally into 'master'?

**Comment by galak** [comment]
I was able to change it to arm

**Comment by galak** [comment]
Close this as I pushed a new version that doesn't have the upstream master commits that I will merge.

**Pull request closed without merging**
